src/fm/cp_dlaf_utils.c

```diff
@@ -0,0 +1,49 @@
+/*----------------------------------------------------------------------------*/
+/*  CP2K: A general program to perform molecular dynamics simulations         */
+/*  Copyright 2000-2023 CP2K developers group <https://cp2k.org>              */
+/*                                                                            */
+/*  SPDX-License-Identifier: GPL-2.0-or-later                                 */
+/*----------------------------------------------------------------------------*/
+
+#if defined(__DLAF)
+
+#include <dlaf_c/eigensolver/eigensolver.h>
+#include <dlaf_c/factorization/cholesky.h>
+#include <dlaf_c/init.h>
+
+void dlaf_init() {
+  const char *pika_argv[] = {"cp2k", "--pika:print-bind"};
+  const char *dlaf_argv[] = {"cp2k"};
+  dlaf_initialize(2, pika_argv, 1, dlaf_argv);
+}
+
+void dlaf_pdsyevd_wrapper(int n, double *a, int ia, int ja, int desca[9],
+                          double *w, double *z, int iz, int jz, int descz[9],
+                          int *info) {
+  char uplo = 'L'; // Only lower triangular matrices are supported
+  dlaf_pdsyevd(uplo, n, a, ia, ja, desca, w, z, iz, jz, descz, info);
+}
+
+void dlaf_pdpotrf_wrapper(int uplo, int n, double *a, int ia, int ja,
+                          int desca[9], int *info) {
+  char _uplo = ' ';
+  if (uplo == 0) {
+    _uplo = 'L';
+  } else if (uplo == 1) {
+    _uplo = 'U';
+  }
```

Comment:
This shows my ignorance with fortran, scalapack, etc. but what is this doing here? From my naive understanding this is always called with `'U'`, so in what situations would you expect to get `0` or `1`?

Reply:
This ugly workaround stems from my Fortran ignorance: it did not manage to pass a `CHARACTER` to C, so I had to resort to numbers.

It is indeed always called with `'U'` for now, but since for the Cholesky decomposition DLA-Future supports both values of UPLO, I wanted to support both.

But given the ugliness of the workaround, probably it is not worth it...

Reply:
Ok, thanks for the explanation! I don't really understand the issue properly so I can't say if there's another way :shrug: So from my perspective this is fine, I was curious since it looked a bit odd. Does @mtaillefumier have some more experience with this? Is there another way of handling this?

Reply:
@msimberg I have a small reproducer for the issue: [fortran_to_c_char](https://github.com/RMeli/fortran-playground/tree/main/fortran_to_c_char) (`CHARACTER(len=*)` behaves as expected, `CHARACTER` not really). Of course I might be doing something wrong, I'm not a Fortran expert. It might be down to different sizes of the types? Unfortunately I couldn't find much information about this (all information I could find was focused on how to pass `CHARACTER(len=*)` to C, which works as expected).

IIRC, using integers instead of characters might have been @mtaillefumier's suggestion (or something coming up in a discussion).

Anyways, it's ugly. Since we only call with `"U"` anyways, I'll follow your suggestion.